py_ecc/bls_api/utils.py

```diff
@@ -0,0 +1,228 @@
+        if y_coordinate is not None:  # Check if quadratic residue found
+            break
+        x_coordinate += FQ2([1, 0])  # Add 1 and try again
+
+    return multiply(
+        G2Uncompressed((x_coordinate, y_coordinate, FQ2([1, 0]))),
+        G2_cofactor,
+    )
+
+
+#
+# G1
+#
+def compress_G1(pt: G1Uncompressed) -> G1Compressed:
+    """
+    A compressed point is a 384-bit integer with the bit order (c_flag, b_flag, a_flag, x),
+    where the c_flag bit is always set to 1,
+    the b_flag bit indicates infinity when set to 1,
+    the a_flag bit helps determine the y-coordinate when decompressing,
+    and the 381-bit integer x is the x-coordinate of the point.
+    """
+    if is_inf(pt):
+        # Set c_flag = 1 and b_flag = 1. leave a_flag = x = 0
+        return G1Compressed(POW_2_383 + POW_2_382)
+    else:
+        x, y = normalize(pt)
+        # Record y's leftmost bit to the a_flag
+        a_flag = (y.n * 2) // q
+        # Set c_flag = 1 and b_flag = 0
+        return G1Compressed(x.n + a_flag * POW_2_381 + POW_2_383)
+
+
+def decompress_G1(z: G1Compressed) -> G1Uncompressed:
+    """
+    Recovers x and y coordinates from the compressed point.
+    """
+    # b_flag == 1 indicates the infinity point
+    b_flag = (z % POW_2_383) // POW_2_382
+    if b_flag == 1:
+        return G1Uncompressed(Z1)
+    x = z % POW_2_381
+
+    # Try solving y coordinate from the equation Y^2 = X^3 + b
+    # using quadratic residue
+    y = pow((x**3 + b.n) % q, (q + 1) // 4, q)
+
+    if pow(y, 2, q) != (x**3 + b.n) % q:
+        raise ValueError(
+            "The given point is not on G1: y**2 = x**3 + b"
+        )
+    # Choose the y whose leftmost bit is equal to the a_flag
+    a_flag = (z % POW_2_382) // POW_2_381
+    if (y * 2) // q != a_flag:
+        y = q - y
+    return G1Uncompressed((FQ(x), FQ(y), FQ(1)))
+
+
+def G1_to_pubkey(pt: G1Uncompressed) -> BLSPubkey:
+    z = compress_G1(pt)
+    return BLSPubkey(z.to_bytes(48, "big"))
+
+
+def pubkey_to_G1(pubkey: BLSPubkey) -> G1Uncompressed:
+    z = big_endian_to_int(pubkey)
+    return decompress_G1(z)
+
+#
+# G2
+#
+
+
+def compress_G2(pt: G2Uncompressed) -> G2Compressed:
+    """
+    The compressed point (z1, z2) has the bit order:
+    z1: (c_flag1, b_flag1, a_flag1, x1)
+    z2: (c_flag2, b_flag2, a_flag2, x2)
+    where
+    - c_flag1 is always set to 1
+    - b_flag1 indicates infinity when set to 1
+    - a_flag1 helps determine the y-coordinate when decompressing,
+    - a_flag2, b_flag2, and c_flag2 are always set to 0
+    """
+    if not is_on_curve(pt, b2):
+        raise ValueError(
+            "The given point is not on the twisted curve over FQ**2"
+        )
+    if is_inf(pt):
+        return G2Compressed((POW_2_383 + POW_2_382, 0))
+    x, y = normalize(pt)
+    # c_flag1 = 1, b_flag1 = 0
+    x_re, x_im = x.coeffs
+    _, y_im = y.coeffs
+    # Record the leftmost bit of y_im to the a_flag1
+    a_flag1 = (y_im * 2) // q
```

Comment:
What if `y_im = 0`, then you need the flag to determine `(y_re * 2) // q`. I think we need both cases, ie. `(y_im * 2) // q` if `y_im > 0` else `(y_re * 2) // q`